### PR TITLE
bpo-22674: signal: add strsignal()

### PR DESCRIPTION
**Pull request by seirl**
I updated @vajrasky 's patch in https://bugs.python.org/issue22674 to rebase it onto master, use the clinic argument parser and improve the docs.

https://bugs.python.org/issue22674

### CODE REVIEW

**Comment by seirl** [comment]
Updated with a NEWS entry.

**Comment by seirl** [comment]
I have absolutely no idea why it doesn't link on windows, I'd really appreciate if someone has any clue as I never use that OS. :P

**Comment by seirl** [comment]
Updated with a special case for Windows. The tests are all passing now, just waiting for a review.

File: Modules/signalmodule.c [hunk 1]
```diff
@@ -504,6 +504,75 @@ signal_getsignal_impl(PyObject *module, int signalnum)
...
+The return values can be such as 'Interrupt', 'Segmentation fault', etc.
+Returns None if the signal is not recognized.
+[clinic start generated code]*/
+
+static PyObject *
+signal_strsignal_impl(PyObject *module, int signalnum)
+/*[clinic end generated code: output=44e12e1e3b666261 input=5369187c5b204b2a]*/
+{
+    int position;
+    const char *res, *pch;
```
Comment (pitrou) [inline Modules/signalmodule.c]: According to [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/functions/strsignal.html), `strsignal` returns non-const `char*`, so let's keep it like that.

File: Modules/signalmodule.c [hunk 2]
```diff
@@ -504,6 +504,75 @@ signal_getsignal_impl(PyObject *module, int signalnum)
...
+        case SIGTERM:
+            res = "Terminated";
+            break;
+        default:
+            Py_RETURN_NONE;
+    }
+#else
+    res = strsignal(signalnum);
+
+    if (res == NULL || strstr(res, "Unknown signal") != NULL)
```
Comment (pitrou) [inline Modules/signalmodule.c]: According to POSIX, you may also want to check for `errno` after the call (and set it to zero before).

File: Modules/signalmodule.c [hunk 3]
```diff
@@ -504,6 +504,75 @@ signal_getsignal_impl(PyObject *module, int signalnum)
...
+      Py_RETURN_NONE;
+
+    /* On Mac, it returns 'Quit: 3', while on Linux it returns 'Quit'. */
```
Comment (pitrou) [inline Modules/signalmodule.c]: Let's not add this kind of post-processing. If someone wants the same return values on all systems, they should simply not use `strsignal` at all (it's not very hard to come up with descriptions for a dozen or less well-known signals).

File: Doc/library/signal.rst [hunk 4]
```diff
@@ -207,6 +207,15 @@ The :mod:`signal` module defines the following functions:
    installed from Python.
 
 
+.. function:: strsignal(signalnum)
+
+   Return the system description of the signal *signalnum*, such as
+   'Interrupt', 'Segmentation fault', etc. Returns :const:`None` if the signal
```
Comment (pitrou) [inline Doc/library/signal.rst]: Please use double quotes: this is English prose, not Python :-)

File: Modules/signalmodule.c [hunk 5]
```diff
@@ -504,6 +504,75 @@ signal_getsignal_impl(PyObject *module, int signalnum)
...
+            res = "Terminated";
+            break;
+        default:
+            Py_RETURN_NONE;
+    }
+#else
+    res = strsignal(signalnum);
+
+    if (res == NULL || strstr(res, "Unknown signal") != NULL)
+      Py_RETURN_NONE;
```
Comment (pitrou) [inline Modules/signalmodule.c]: Note that we indent C code by 4-space indentations, not 2. See [PEP 7](https://www.python.org/dev/peps/pep-0007/) for more details.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for posting this PR! I've added some comments to the code, please make the necessary changes :-)

File: Modules/signalmodule.c (hunk 1 above)
Comment (seirl) [reply Modules/signalmodule.c]: I don't understand this one. Why would we want to keep it not const? We're not planning on modifying it, adding constness to make sure of that doesn't hurt.

Comment (pitrou) [reply Modules/signalmodule.c]: `Py_BuildValue` doesn't specify that it takes only `const` arguments, so you may get compiler warnings or even errors depending on compiler settings. The benefit of using `const` in such a simple function seem rather low, so it's better to drop it iMHO.

**Comment by seirl** [comment]
I have made the requested changes; please review again.

File: Modules/signalmodule.c (hunk 1 above)
Comment (seirl) [reply Modules/signalmodule.c]: Got it, I just looked at the docs of Py_Buildvalue. I fixed it, thanks.

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
Thank you @seirl ! This looks good to me now.

**Pull request merged**